### ISSUE TITLE
[BUG] Uploading 1GB model takes 5 minutes, with high CPU/memory usage

### ISSUE THREAD

**Issue by hanlaur** [mlflow/mlflow#22665]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.11.1
- Tracking server: 3.11.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS 26.4.1
- **Python version**: 3.12

### Describe the problem

This is a reopen of https://github.com/mlflow/mlflow/issues/18696, verified against latest mlflow version.

When I upload 1GB file against mlflow server on localhost, it takes >5 minutes and mlflow memory consumption climbs to 30+ GB on my laptop (+heavy GC activity and heavy CPU usage)

While PR https://github.com/mlflow/mlflow/pull/19520 fixed the artifact download issue https://github.com/mlflow/mlflow/issues/19517 nicely, the model upload performance regression https://github.com/mlflow/mlflow/issues/18696 does not seem to be resolved.

The performance problem seems to be caused by the WSGIMiddleware, specifically this line:
https://github.com/Kludex/starlette/blob/edcc79eb6df0319c25aeeae77378a0220facf181/starlette/middleware/wsgi.py#L103

The bytes concatenation gets more and more expensive as the data size grows. In my debugging I observed data was received 256kB at a time. With a surgical patch to the WSGIMiddleware while loop to collect byte chunks first to a list and only join them after all chunks have been collected to contruct the full byte array, the performance improves dramatically and upload completes in 4 seconds.

However, appears the whole WSGIMiddleware is deprecated, so it may be that proposing patch to currently used WSGIMiddleware it is not the way to go:
https://github.com/Kludex/starlette/blob/edcc79eb6df0319c25aeeae77378a0220facf181/starlette/middleware/wsgi.py#L15 . Probably potential fix could be either to try the suggested alternative WSGIMiddleware, fork the WSGIMiddleware or complete the migration out of WSGIMiddleware assuming it is meant only as temporary approach in mlflow.

### Tracking information

```shell
System information: Darwin Darwin Kernel Version 25.4.0: Thu Mar 19 19:33:25 PDT 2026; root:xnu-12377.101.15~1/RELEASE_ARM64_T6041
Python version: 3.12.13
MLflow version: 3.11.1
Tracking URI: http://localhost:5000
Registry URI: http://localhost:5000
MLflow environment variables: 
  MLFLOW_TRACKING_URI: http://localhost:5000
MLflow dependencies: 
  Flask: 3.1.3
  Flask-CORS: 6.0.2
  Flask-WTF: 1.2.2
  aiohttp: 3.13.5
  alembic: 1.18.4
  boto3: 1.42.91
  botocore: 1.42.91
  click: 8.3.2
  cryptography: 46.0.7
  docker: 7.1.0
  fastapi: 0.136.0
  graphene: 3.4.3
  gunicorn: 25.3.0
  huey: 2.6.0
  kubernetes: 35.0.0
  matplotlib: 3.10.8
  mlflow-skinny: 3.11.1
  mlflow-tracing: 3.11.1
  numpy: 2.4.4
  pandas: 2.3.3
  psycopg2-binary: 2.9.11
  pyarrow: 23.0.1
  scikit-learn: 1.8.0
  scipy: 1.17.1
  skops: 0.13.0
  sqlalchemy: 2.0.49
  uvicorn: 0.44.0
```

### Code to reproduce issue

```python
import time

import mlflow
import mlflow.sklearn
import numpy as np
from sklearn.base import BaseEstimator, RegressorMixin
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler


class DummyLargeRegressor(BaseEstimator, RegressorMixin):
    def fit(self, X, y=None):
        self.random_payload_ = np.random.randint(0, 256, size=1000 * 1024 * 1024, dtype=np.uint8)
        return self

    def predict(self, X):
        return np.zeros(len(X))


def main():
    mlflow.set_tracking_uri("http://localhost:5000")

    X_train = np.random.randn(100, 10)
    y_train = np.random.randn(100)

    pipeline = Pipeline(
        [
            ("scaler", StandardScaler()),
            ("regressor", DummyLargeRegressor()),
        ]
    )

    pipeline.fit(X_train, y_train)

    print("Logging model to MLflow...")
    with mlflow.start_run() as run:
        start_time = time.time()
        mlflow.sklearn.log_model(pipeline, registered_model_name="registered")
        upload_time = time.time() - start_time
        run_id = run.info.run_id
        print(f"Model logged with run_id: {run_id}")
        print(f"Model upload took {upload_time:.2f} seconds")

    print("Downloading model from MLflow...")
    client = mlflow.MlflowClient()
    latest_version_info = client.search_model_versions(f"run_id='{run_id}'")[0]
    start_time = time.time()
    loaded_model = mlflow.sklearn.load_model(latest_version_info.source)
    download_time = time.time() - start_time
    print(f"Model download took {download_time:.2f} seconds")

    regressor = loaded_model.named_steps["regressor"]
    payload_size_mb = len(regressor.random_payload_) / (1024 * 1024)
    print(f"Downloaded model payload size: {payload_size_mb:.2f} MB")

if __name__ == "__main__":
    main()

```

### Stack trace

```
N/A
```

### Other info / logs

```
N/A
```

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by harupy** [comment]
## Root Cause

Starlette's `WSGIMiddleware` ([`wsgi.py:97-101`](https://github.com/Kludex/starlette/blob/edcc79eb6df0319c25aeeae77378a0220facf181/starlette/middleware/wsgi.py#L97-L101)) buffers the entire request body using `body += chunk` in a loop — O(n^2) for large payloads. MLflow wraps Flask inside FastAPI via this middleware ([`fastapi_app.py:142`](https://github.com/mlflow/mlflow/blob/master/mlflow/server/fastapi_app.py#L142)), so all artifact uploads go through it.

Fix: #22729

**Comment by harupy** [comment]
## Testing the fix

```bash
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/22729/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/22729/merge#subdirectory=libs/skinny
```
